File: output/GPT4o-Mini/xCodeEval_4000/C++/Go/cd2a9d8be7b95e06801b5c5d33034163.go
package main

import (
	"fmt"
	"math"
	"sort"
)

type DSum struct {
	N, K int
	A    [][]int64
	A_sum []int64
	best  int64
}

func (ds *DSum) addItem(knapsack []int64, size int, sum int64) {
	for k := ds.K - size; k >= 0; k-- {
		knapsack[k+size] = max(knapsack[k+size], knapsack[k]+sum)
	}
}

func (ds *DSum) recurse(start, end int, knapsack []int64) {
	if start >= end {
		return
	}

	if end-start == 1 {
		sum := int64(0)
		ds.best = max(ds.best, knapsack[ds.K])

		for prefix := 1; prefix <= len(ds.A[start]); prefix++ {
			sum += ds.A[start][prefix-1]
			ds.best = max(ds.best, sum+knapsack[ds.K-prefix])
		}

		return
	}

	mid := (start + end) / 2
	state := make([]int64, len(knapsack))
	copy(state, knapsack)

	for i := start; i < mid; i++ {
		ds.addItem(state, len(ds.A[i]), ds.A_sum[i])
	}

	ds.recurse(mid, end, state)
	state = make([]int64, len(knapsack))
	copy(state, knapsack)

	for i := mid; i < end; i++ {
		ds.addItem(state, len(ds.A[i]), ds.A_sum[i])
	}

	ds.recurse(start, mid, state)
}

func (ds *DSum) solve() {
	fmt.Scan(&ds.N, &ds.K)
	ds.A = make([][]int64, ds.N)
	ds.A_sum = make([]int64, ds.N)

	for i := 0; i < ds.N; i++ {
		var t int
		fmt.Scan(&t)
		ds.A[i] = make([]int64, t)

		for j := 0; j < t; j++ {
			fmt.Scan(&ds.A[i][j])
		}

		if t > ds.K {
			ds.A[i] = ds.A[i][:ds.K]
		}

		ds.A_sum[i] = sum(ds.A[i])
	}

	ds.best = 0
	knapsack := make([]int64, ds.K+1)
	ds.recurse(0, ds.N, knapsack)
	fmt.Println(ds.best)
}

func sum(arr []int64) int64 {
	total := int64(0)
	for _, v := range arr {
		total += v
	}
	return total
}

func max(a, b int64) int64 {
	if a > b {
		return a
	}
	return b
}

func main() {
	var solver DSum
	solver.solve()
}

File: output/GPT4o-Mini/xCodeEval_remove_4000/C++/Python/cd2a9d8be7b95e06801b5c5d33034163.py
import sys
from itertools import accumulate

class DSum:
    class Solver:
        def __init__(self, in_stream, out_stream):
            self._in = in_stream
            self._out = out_stream
            self.N = 0
            self.K = 0
            self.A = []
            self.A_sum = []
            self.best = 0

        def rd(self):
            return map(int, self._in.readline().split())

        def pr(self, *args):
            self._out.write(' '.join(map(str, args)))

        def prln(self, *args):
            self.pr(*args)
            self._out.write('\n')

        def add_item(self, knapsack, size, sum_):
            for k in range(self.K - size, -1, -1):
                knapsack[k + size] = max(knapsack[k + size], knapsack[k] + sum_)

        def recurse(self, start, end, knapsack):
            if start >= end:
                return

            if end - start == 1:
                sum_ = 0
                self.best = max(self.best, knapsack[self.K])

                for prefix in range(1, len(self.A[start]) + 1):
                    sum_ += self.A[start][prefix - 1]
                    self.best = max(self.best, sum_ + knapsack[self.K - prefix])

                return

            mid = (start + end) // 2
            state = knapsack[:]

            for i in range(start, mid):
                self.add_item(state, len(self.A[i]), self.A_sum[i])

            self.recurse(mid, end, state)
            state = knapsack[:]

            for i in range(mid, end):
                self.add_item(state, len(self.A[i]), self.A_sum[i])

            self.recurse(start, mid, state)

        def solve(self):
            self.N, self.K = self.rd()
            self.A = [[] for _ in range(self.N)]
            self.A_sum = [0] * self.N

            for i in range(self.N):
                t = int(self._in.readline())
                self.A[i] = list(map(int, self._in.readline().split()))

                if t > self.K:
                    self.A[i] = self.A[i][:self.K]

                self.A_sum[i] = sum(self.A[i])

            self.best = 0
            knapsack = [0] * (self.K + 1)
            self.recurse(0, self.N, knapsack)
            self.prln(self.best)

    @staticmethod
    def solve(in_stream, out_stream):
        DSum.Solver(in_stream, out_stream).solve()

if __name__ == "__main__":
    DSum.solve(sys.stdin, sys.stdout)